include/zio/zio_attr.h

```diff
@@ -0,0 +1,179 @@
+
+/**
+ * @file
+ * @brief ZIO attribute definitions.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_ATTR_H_
+#define ZEPHYR_INCLUDE_ZIO_ATTR_H_
+
+#include <zephyr/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief ZIO attribute enum and struct definitions.
+ * @defgroup zio_attributes ZIO attribute definitions
+ * @ingroup zio
+ * @{
+ */
+
+/**
+ * @brief A type tag enum
+ */
+/**< Value type identifier. */
+enum zio_attr_type {
+	zio_attr_float,                 /**< Single-precision float tag. */
+	zio_attr_double,                /**< Double-precision float tag. */
+	zio_attr_bool,                  /**< 1-bit boolean tag. */
+	zio_attr_u8,                    /**< Unsigned 8-bit integer tag. */
+	zio_attr_u16,                   /**< Unsigned 16-bit integer tag. */
+	zio_attr_u32,                   /**< Unsigned 32-bit integer tag. */
+	zio_attr_u64,                   /**< Unsigned 64-bit integer tag. */
+	zio_attr_s8,                    /**< Signed 8-bit integer tag. */
+	zio_attr_s16,                   /**< Signed 16-bit integer tag. */
+	zio_attr_s32,                   /**< Signed 32-bit integer tag. */
+	zio_attr_s64,                   /**< Signed 64-bit integer tag. */
+	zio_attr_str,                   /**< String tag. */
+	zio_attr_ptr,                   /**< Generic pointer tag. */
+};
+
+/**
+ * @brief Tag for a value of known type
+ */
+ #define zio_attr_tag(val) _Generic((val),			      \
+				    float : zio_attr_float,	      \
+				    double : zio_attr_double,	      \
+				    bool : zio_attr_bool,	      \
+				    unsigned char : zio_attr_u8,      \
+				    unsigned short : zio_attr_u16,    \
+				    unsigned long : zio_attr_u32,     \
+				    unsigned long long, zio_attr_u64, \
+				    signed char : zio_attr_s8,	      \
+				    signed short : zio_attr_s16,      \
+				    signed long : zio_attr_s32,	      \
+				    signed long long : zio_attr_s64,  \
+				    char * : zio_attr_str	      \
+				    void * : zio_attr_ptr	      \
+				    )
+
+/**
+ * @brief A type tagged variant for attributes
+ */
+struct zio_attr_data {
+	enum zio_attr_type tag;
+
+	/**< Assigned value. */
+	union {
+		float float_val;        /**< Single-precision float value. */
+		double double_val;      /**< Double-precision float value. */
+		u8_t bool_val : 1;      /**< 1-bit boolean value. */
+		u8_t u8_val;            /**< Unsigned 8-bit integer value. */
+		u16_t u16_val;          /**< Unsigned 16-bit integer value. */
+		u32_t u32_val;          /**< Unsigned 32-bit integer value. */
+		u64_t u64_val;          /**< Unsigned 64-bit integer value. */
+		s8_t s8_val;            /**< Signed 8-bit integer value. */
+		s16_t s16_val;          /**< Signed 16-bit integer value. */
+		s32_t s32_val;          /**< Signed 32-bit integer value. */
+		s64_t s64_val;          /**< Signed 64-bit integer value. */
+		char  *str_val;         /**< String (char *) value. */
+		void  *ptr_val;         /**< Generic pointer (void *) value. */
```

Comment:
No, the format of the data is defined by the device driver and described by an array of channel description structs. It is a goal to avoid do any data conversion unless an application calls functions to do so.

Reply:
So this type is just for attributes, not for data samples?